### PR TITLE
Update "tests/README.md"

### PR DESCRIPTION
**Pull request by arshad512**
This patch adds detail section on adding and running
test-case. It also changes markdown number list to
more readeable headers

Signed-off-by: Arshad Hussain <arshad.hussain@aeoncomputing.com>

This patch updates and improve tests/README.md to add a section on how to add new test case.

### Motivation and Context

Details was lacking.

### How Has This Been Tested?
Yes it is being tested.

This is tested by 'grip' tool for any markdown mismatch.

Tested using 'grip' tool on vm

N/A

N/A

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: tests/README.md [hunk 1]
```diff
@@ -146,7 +146,98 @@ with the `zfs-tests.sh` wrapper script will look something like this:
...
+
     Running Time:	02:35:33
     Percent passed:	95.6%
     Log directory:	/var/tmp/test_results/20180515T054509
+
+### 4) Example of adding and running test-case (zpool_example)
+
+  This broadly boils down to 5 steps
+  1. Create/Set password-less sudo for user running test case.
+  2. Edit configure.ac, Makefile.ac appropriately
```
Comment (freqlabs) [inline tests/README.md:158]:
```suggestion
  2. Edit configure.ac, Makefile.am appropriately
```

**Comment by freqlabs** [comment]
I like the direction of this. It is probably worth mentioning up front that usually we're not adding new runfiles when adding test cases, but rather modifying an existing runfile. Having the process of adding a new runfile documented as well still is a good thing, IMO, so maybe a section describing the existing runfiles will help people choose the correct place to make their additions.

**Comment by arshad512** [comment]
> I like the direction of this. It is probably worth mentioning up front that usually we're not adding new runfiles when adding test cases, but rather modifying an existing runfile. Having the process of adding a new runfile documented as well still is a good thing, IMO. Maybe a section describing the existing runfiles will help people choose the correct place to make their additions.

Thanks for the review. I will update section to include section for updating the runfile. The 'zpool_example' what I have updated is about adding altogether the new test-case section. For that adding new runfile is necessary IMO. That said, I will try to incorporate both section of adding new runfile and updating the existing runfiles (as you have ponted out).

**Review by tonynguien — APPROVED** [review]
Thanks for adding the steps for new tests. I'm requesting @jwk404 to make a pass at this.

**Comment by tonynguien** [comment]
@arshad512 Can you add the sign-off message to the commit, e.g. `git commit --all --signoff` to resolve the checkstyle error?

See this page for OpenZFS workflow: 
https://openzfs.github.io/openzfs-docs/Developer%20Resources/Git%20and%20GitHub%20for%20beginners.html#committing-your-changes-to-be-pushed

Don't hesitate to reach out if you have questions.

**Comment by jwk404** [comment]
> I will update section to include section for updating the runfile. The 'zpool_example' what I have updated is about adding altogether the new test-case section. For that adding new runfile is necessary IMO. That said, I will try to incorporate both section of adding new runfile and updating the existing runfiles (as you have ponted out).

I don't see these changes about the runfiles - is it possible they're in another commit that's yet to be pushed?

**Comment by arshad512** [comment]
> > I will update section to include section for updating the runfile. The 'zpool_example' what I have updated is about adding altogether the new test-case section. For that adding new runfile is necessary IMO. That said, I will try to incorporate both section of adding new runfile and updating the existing runfiles (as you have ponted out).
> 
> I don't see these changes about the runfiles - is it possible they're in another commit that's yet to be pushed?

Hi @jwk404 , @tonynguien 

Apologies for late reply. There is no another push. I could not complete my changes as I was pulled into family emergency. I should be updating the writeup in this week.

**Comment by arshad512** [comment]
> @arshad512 Can you add the sign-off message to the commit, e.g. `git commit --all --signoff` to resolve the checkstyle error?
> 
> See this page for OpenZFS workflow: https://openzfs.github.io/openzfs-docs/Developer%20Resources/Git%20and%20GitHub%20for%20beginners.html#committing-your-changes-to-be-pushed
> 
> Don't hesitate to reach out if you have questions.

@tonynguien , Sorry, I missed your comment earlier. I do have a "Signed-off-by" line in commit log.  This time around I am just doing "git commit --amend" and force-pushing as I am just updating a small seciton of file. I have revised the redme and pushed a new version. Please review/check if this fixes what you want.

**Comment by tonynguien** [comment]
> > @arshad512 Can you add the sign-off message to the commit, e.g. `git commit --all --signoff` to resolve the checkstyle error?
> > See this page for OpenZFS workflow: https://openzfs.github.io/openzfs-docs/Developer%20Resources/Git%20and%20GitHub%20for%20beginners.html#committing-your-changes-to-be-pushed
> > Don't hesitate to reach out if you have questions.
> 
> @tonynguien , Sorry, I missed your comment earlier. I do have a "Signed-off-by" line in commit log. This time around I am just doing "git commit --amend" and force-pushing as I am just updating a small seciton of file. I have revised the redme and pushed a new version. Please review/check if this fixes what you want.

Thanks! It looks good to me.

**Review by tonynguien — APPROVED** [review]

**Review by jwk404 — APPROVED** [review]

File: tests/README.md (hunk 1 above)
Comment (freqlabs) [reply tests/README.md:158]: Please fix

File: tests/README.md [hunk 2]
```diff
@@ -146,7 +146,111 @@ with the `zfs-tests.sh` wrapper script will look something like this:
...
+  3. Create/Modify .run files
+  4. Create actual test-scripts
+  5. Run Test case
+
+  Will look at each of them in depth.
+
+  * Set password-less sudo for 'Test' user as test script cannot be run as root
+  * Edit file **configure.am** and include line under AC_CONFIG_FILES section
```
Comment (freqlabs) [inline tests/README.md:166]:
```suggestion
  * Edit file **configure.ac** and include line under AC_CONFIG_FILES section
```

File: tests/README.md [hunk 3]
```diff
@@ -146,7 +146,111 @@ with the `zfs-tests.sh` wrapper script will look something like this:
...
+	# Prep backend device
+	log_must dd if=/dev/zero of=$DISKS_DEV1 bs=512 count=140000
+	log_must dd if=/dev/zero of=$DISKS_DEV2 bs=512 count=140000
+
+	# Create pool
+	log_must zpool create $TESTPOOL $type $DISKS_DEV1 $DISKS_DEV2
+
+	log_pass "zpool_example"
+    ~~~~
+  * Run Test case. Which can be done in two ways. Described in detail above in
```
Comment (freqlabs) [inline tests/README.md:253]:
```suggestion
  * Run Test case, which can be done in two ways. Described in detail above in
```

**Review by freqlabs — CHANGES_REQUESTED** [review]

File: tests/README.md (hunk 2 above)
Comment (arshad512) [reply tests/README.md]: Thanks for review! Resolved in new push.

File: tests/README.md (hunk 3 above)
Comment (arshad512) [reply tests/README.md]: Resolved in new push. Thanks

**Comment by freqlabs** [comment]
My comment on line 158 (Makefile.ac -> Makefile.am) is still not resolved.

**Comment by arshad512** [comment]
> My comment on line 158 (Makefile.ac -> Makefile.am) is still not resolved.

I do not know how I missed this. Sorry for that. This is fixed now.

**Review by behlendorf — APPROVED** [review]

File: tests/README.md [hunk 4]
```diff
@@ -146,7 +146,111 @@ with the `zfs-tests.sh` wrapper script will look something like this:
...
+        common.run \
+        freebsd.run \
+        linux.run \
+        longevity.run \
+        perf-regression.run \
+        sanity.run \
+        sunos.run
+    ~~~~
+  * Create file **tests/runfiles/zpool_example.run**. This defines the most
+    common properties when run with test-runner.py or zfs-test.sh.
```
Comment (freqlabs) [inline tests/README.md:184]:
```suggestion
    common properties when run with test-runner.py or zfs-tests.sh.
```

File: tests/README.md [hunk 5]
```diff
@@ -146,7 +146,111 @@ with the `zfs-tests.sh` wrapper script will look something like this:
...
+	{
+		# Cleanup
+		destroy_pool $TESTPOOL
+		log_must rm -f $DISKS_DEV1
+		log_must rm -f $DISKS_DEV2
+	}
+
+	# Run function "cleanup" on exit
+	log_onexit cleanup
+
```
Comment (freqlabs) [inline tests/README.md:243]: Test cases should also `log_assert "Something."` somewhere in the opening.

**Review by freqlabs — CHANGES_REQUESTED** [review]

File: tests/README.md (hunk 4 above)
Comment (arshad512) [reply tests/README.md:184]: Yes. I will update this over the weekend.

File: tests/README.md (hunk 5 above)
Comment (arshad512) [reply tests/README.md:243]: I am not sure if this is required for the given sample.  However, I will go through the example and incorporate your suggestion.

Comment (arshad512) [reply tests/README.md:244]: Done. Thanks! I see what you meant +1:

File: tests/README.md (hunk 4 above)
Comment (arshad512) [reply tests/README.md]: Done

File: tests/README.md [hunk 6]
```diff
@@ -146,7 +146,112 @@ with the `zfs-tests.sh` wrapper script will look something like this:
...
+      [DEFAULT]
+      timeout = 600
+      outputdir = /var/tmp/test_results
+      tags = ['functional']
+
+      tests = ['zpool_example_001_pos']
+    ~~~~
+    If adding test-case to an already existing suite the runfile would
+    already be present and it needs to be only updated. For example, adding
+    **zpool_example_002_pos** to the above runfile only update the **"test ="**
```
Comment (freqlabs) [inline tests/README.md:195]:
```suggestion
    **zpool_example_002_pos** to the above runfile only update the **"tests ="**
```

**Review by freqlabs — COMMENTED** [review]
I see one more error. Thank you for being so patient!

File: tests/README.md (hunk 6 above)
Comment (arshad512) [reply tests/README.md]: Done. Thanks for spotting this :+1:

**Review by freqlabs — APPROVED** [review]

**Pull request merged**